### DISCUSSION TITLE
Airflow memory overloads and everything stales

### DISCUSSION THREAD

**Discussion by ivanrezic** [apache/airflow#17857] [Q&A]
Hello everyone,

I am developing a project using Airflow and I am having some issues. I have tried googling around but it didn't help. I think my problem is in conceptual understanding of Airflow so I am quite not sure where to start. I am hoping you could point me in the right direction after I describe what I want and what I am struggling with.

First of all, project requirements are simple:

- connect to the database and fetch N listings (MongoDB documents)

- extract listings image paths and load images

- process and upload belonging images to the destination

I should do these 3 steps as fast as possible on 8 core machine with ~19 GB of spare RAM. One iteration through these steps takes ~2 mins.

As I want to do it **as fast as possible** I have set `schedule_interval="* * * * *"` or in other words to **run it every minute**. Also, I want to run it using multiple processes so I am duplicating these 3 steps multiple times (for example 3 times). What I get is this form of pipeline **Start >> [Steps1, Steps2, Steps3] >> End** where Start and END are just dummy operators to connect running steps. And finally, I have set `max_active_runs=1` to run only one dag at a time.

Here is my code without util methods, I think they are nothing important as they are simple everyday operations.

```
DEFAULT_ARGS = {
    "owner": "airflow",
    "retries": 30,
    "retry_delay": timedelta(minutes=5),
    "depends_on_past": False,
}

dag = DAG(
    dag_id="mongo_process_upload",
    start_date=datetime(2021, 3, 5, 10, 30),
    schedule_interval="* * * * *",
    default_args=DEFAULT_ARGS,
    max_active_runs=1,
    catchup=False,
)

start = DummyOperator(task_id='start', dag=dag)
end = DummyOperator(task_id='end', dag=dag)

db = get_mongo_db().get_database(CF["base"])

collection_spider = db[CF["spider"]]
collection_send = db[CF["destination"]]

cursor_spider = (
    collection_spider.find(CF["spider_filter"], no_cursor_timeout=True)
        .limit(CF["number"])
        .skip(CF["offset"])
        .sort('updated_at', ASCENDING)
)

for document in cursor_spider:
    steps = PythonOperator(
        task_id=PUBLISH_TASK_ID.format(document['_id']),
        python_callable=publish_pics,
        provide_context=True,
        op_kwargs={"doc": document},
        dag=dag,
    )

    start >> steps >> end
```

After running it, everything **seems fine for a few minutes** but then, memory reaches 100% of usage and everything slows down and stales. I need to restart airflow manually to start processing again.

And here are my questions. 

Am I doing something conceptually wrong here? Why does memory overload happen? I have a feeling that somehow Airflow does something in the background I am not aware of? Does it somehow start more processes than wanted or something similar? 

Is Airflow supposed to be used like this? Should I fetch listings each minute and create a new pipeline with new 3 tasks or should I do it somehow different? 

Any help or suggestion is greatly appreciated. 
Thank you all in advance!

**Comment by blcksrx** [comment]
In my opinion, you are pressuring airflow with a huge amount of data that led you to this situation.
Even transferring the larger amount of data using TrnasformsOperator could lead you to this situation.
Try to use a smaller batch of data and tasks in every dag run or you can outsource it from airflow. For example, use spark and SaprkSubmitOperator

**Comment by antiny** [comment]
Can you also share your airflow setup? When you said memory reached 100%, do you mean the scheduler or the web server? The recent published book `Data Pipelines with Apache Airflow` is an excellent reference to understand the inner working of Airflow.
